crates/storage/libmdbx-rs/src/transaction.rs

```diff
@@ -88,35 +88,43 @@ where
     }
 
     pub(crate) fn new_from_ptr(env: &'env Environment<E>, txn: *mut ffi::MDBX_txn) -> Self {
-        Self {
-            txn: Rc::new(Mutex::new(txn)),
+        let inner = TransactionInner {
+            txn: TransactionPtr::new(txn),
             primed_dbis: Mutex::new(IndexSet::new()),
-            committed: false,
+            committed: AtomicBool::new(false),
             env,
             _marker: PhantomData,
-        }
+        };
+        Self { inner: Arc::new(inner) }
     }
 
-    /// Returns a raw pointer to the underlying MDBX transaction.
+    /// Executes the given closure once the lock on the transaction is acquired.
     ///
     /// The caller **must** ensure that the pointer is not used after the
     /// lifetime of the transaction.
-    pub(crate) fn txn_mutex(&self) -> Rc<Mutex<*mut ffi::MDBX_txn>> {
-        self.txn.clone()
+    #[inline]
+    pub(crate) fn txn_execute<F: FnOnce(*mut ffi::MDBX_txn) -> T, T>(&self, f: F) -> T {
+        self.inner.txn_execute(f)
+    }
+
+    pub(crate) fn txn_ptr(&self) -> TransactionPtr {
+        self.inner.txn.clone()
     }
 
+    /// Returns a copy of the raw pointer to the underlying MDBX transaction.
+    #[doc(hidden)]
     pub fn txn(&self) -> *mut ffi::MDBX_txn {
```

Comment:
This should probably be removed no? Since it's out of the mutex

Reply:
ideally yes, we currently use this in bench code, will phase out separately